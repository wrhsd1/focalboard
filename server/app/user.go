package app

import (
	"github.com/mattermost/focalboard/server/model"
	mmModel "github.com/mattermost/mattermost-server/v6/model"
)

func (a *App) GetTeamUsers(teamID string) ([]*model.User, error) {
	return a.store.GetUsersByTeam(teamID)
}

func (a *App) SearchTeamUsers(teamID string, searchQuery string) ([]*model.User, error) {
	return a.store.SearchUsersByTeam(teamID, searchQuery)
}

func (a *App) UpdateUserConfig(userID string, patch model.UserPropPatch) (map[string]interface{}, error) {
	if err := a.store.PatchUserProps(userID, patch); err != nil {
		return nil, err
	}

	user, err := a.store.GetUserByID(userID)
	if err != nil {
		return nil, err
	}

	return user.Props, nil
}

func (a *App) SearchUserChannels(teamID string, userID string, query string) ([]*mmModel.Channel, error) {
	return a.store.SearchUserChannels(teamID, userID, query)
}

func (a *App) GetChannel(teamID string, channelID string) (*mmModel.Channel, error) {
	return a.store.GetChannel(teamID, channelID)
}
