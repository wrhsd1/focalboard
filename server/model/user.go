package model

import (
	"encoding/json"
	"io"
)

const (
	SingleUser   = "single-user"
	GlobalTeamID = "0"
	SystemUserID = "system"
)

// User is a user
// swagger:model
type User struct {
	// The user ID
	// required: true
	ID string `json:"id"`

	// The user name
	// required: true
	Username string `json:"username"`

	// The user's email
	// required: true
	Email string `json:"-"`

	// The user's nickname
	Nickname string `json:"nickname"`
	// The user's first name
	FirstName string `json:"firstname"`
	// The user's last name
	LastName string `json:"lastname"`

	// swagger:ignore
	Password string `json:"-"`

	// swagger:ignore
	MfaSecret string `json:"-"`

	// swagger:ignore
	AuthService string `json:"-"`

	// swagger:ignore
	AuthData string `json:"-"`

	// User settings
	// required: true
	Props map[string]interface{} `json:"props"`

	// Created time in miliseconds since the current epoch
	// required: true
	CreateAt int64 `json:"create_at,omitempty"`

	// Updated time in miliseconds since the current epoch
	// required: true
	UpdateAt int64 `json:"update_at,omitempty"`

	// Deleted time in miliseconds since the current epoch, set to indicate user is deleted
	// required: true
	DeleteAt int64 `json:"delete_at"`

	// If the user is a bot or not
	// required: true
	IsBot bool `json:"is_bot"`

	// If the user is a guest or not
	// required: true
	IsGuest bool `json:"is_guest"`

	Roles string `json:"roles"`
}

// UserPropPatch is a user property patch
// swagger:model
type UserPropPatch struct {
	// The user prop updated fields
	// required: false
	UpdatedFields map[string]string `json:"updatedFields"`

	// The user prop removed fields
	// required: false
	DeletedFields []string `json:"deletedFields"`
}

type Session struct {
	ID          string                 `json:"id"`
	Token       string                 `json:"token"`
	UserID      string                 `json:"user_id"`
	AuthService string                 `json:"authService"`
	Props       map[string]interface{} `json:"props"`
	CreateAt    int64                  `json:"create_at,omitempty"`
	UpdateAt    int64                  `json:"update_at,omitempty"`
}

func UserFromJSON(data io.Reader) (*User, error) {
	var user User
	if err := json.NewDecoder(data).Decode(&user); err != nil {
		return nil, err
	}
	return &user, nil
}
