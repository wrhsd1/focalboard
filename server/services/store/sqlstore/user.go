package sqlstore

import (
	"database/sql"
	"encoding/json"
	"errors"
	"fmt"

	sq "github.com/Masterminds/squirrel"

	"github.com/mattermost/focalboard/server/model"
	"github.com/mattermost/focalboard/server/utils"

	"github.com/mattermost/mattermost-server/v6/shared/mlog"
)

var (
	errUnsupportedOperation = errors.New("unsupported operation")
)

type UserNotFoundError struct {
	id string
}

func (unf UserNotFoundError) Error() string {
	return fmt.Sprintf("user not found (%s)", unf.id)
}

func (s *SQLStore) getRegisteredUserCount(db sq.BaseRunner) (int, error) {
	query := s.getQueryBuilder(db).
		Select("count(*)").
		From(s.tablePrefix + "users").
		Where(sq.Eq{"delete_at": 0})
	row := query.QueryRow()

	var count int
	err := row.Scan(&count)
	if err != nil {
		return 0, err
	}

	return count, nil
}

func (s *SQLStore) getUserByCondition(db sq.BaseRunner, condition sq.Eq) (*model.User, error) {
	users, err := s.getUsersByCondition(db, condition, 0)
	if err != nil {
		return nil, err
	}

	if len(users) == 0 {
		return nil, nil
	}

	return users[0], nil
}

func (s *SQLStore) getUsersByCondition(db sq.BaseRunner, condition interface{}, limit uint64) ([]*model.User, error) {
	query := s.getQueryBuilder(db).
		Select(
			"id",
			"username",
			"email",
			"password",
			"mfa_secret",
			"auth_service",
			"auth_data",
			"props",
			"create_at",
			"update_at",
			"delete_at",
		).
		From(s.tablePrefix + "users").
		Where(sq.Eq{"delete_at": 0}).
		Where(condition)

	if limit != 0 {
		query = query.Limit(limit)
	}

	rows, err := query.Query()
	if err != nil {
		s.logger.Error(`getUsersByCondition ERROR`, mlog.Err(err))
		return nil, err
	}
	defer s.CloseRows(rows)

	users, err := s.usersFromRows(rows)
	if err != nil {
		return nil, err
	}

	if len(users) == 0 {
		return nil, sql.ErrNoRows
	}

	return users, nil
}

func (s *SQLStore) getUserByID(db sq.BaseRunner, userID string) (*model.User, error) {
	return s.getUserByCondition(db, sq.Eq{"id": userID})
}

func (s *SQLStore) getUsersList(db sq.BaseRunner, userIDs []string) ([]*model.User, error) {
	return s.getUsersByCondition(db, sq.Eq{"id": userIDs}, 0)
}

func (s *SQLStore) getUserByEmail(db sq.BaseRunner, email string) (*model.User, error) {
	return s.getUserByCondition(db, sq.Eq{"email": email})
}

func (s *SQLStore) getUserByUsername(db sq.BaseRunner, username string) (*model.User, error) {
	return s.getUserByCondition(db, sq.Eq{"username": username})
}

func (s *SQLStore) createUser(db sq.BaseRunner, user *model.User) error {
	now := utils.GetMillis()

	propsBytes, err := json.Marshal(user.Props)
	if err != nil {
		return err
	}

	query := s.getQueryBuilder(db).Insert(s.tablePrefix+"users").
		Columns("id", "username", "email", "password", "mfa_secret", "auth_service", "auth_data", "props", "create_at", "update_at", "delete_at").
		Values(user.ID, user.Username, user.Email, user.Password, user.MfaSecret, user.AuthService, user.AuthData, propsBytes, now, now, 0)

	_, err = query.Exec()
	return err
}

func (s *SQLStore) updateUser(db sq.BaseRunner, user *model.User) error {
	now := utils.GetMillis()

	propsBytes, err := json.Marshal(user.Props)
	if err != nil {
		return err
	}

	query := s.getQueryBuilder(db).Update(s.tablePrefix+"users").
		Set("username", user.Username).
		Set("email", user.Email).
		Set("props", propsBytes).
		Set("update_at", now).
		Where(sq.Eq{"id": user.ID})

	result, err := query.Exec()
	if err != nil {
		return err
	}

	rowCount, err := result.RowsAffected()
	if err != nil {
		return err
	}

	if rowCount < 1 {
		return UserNotFoundError{user.ID}
	}

	return nil
}

func (s *SQLStore) updateUserPassword(db sq.BaseRunner, username, password string) error {
	now := utils.GetMillis()

	query := s.getQueryBuilder(db).Update(s.tablePrefix+"users").
		Set("password", password).
		Set("update_at", now).
		Where(sq.Eq{"username": username})

	result, err := query.Exec()
	if err != nil {
		return err
	}

	rowCount, err := result.RowsAffected()
	if err != nil {
		return err
	}

	if rowCount < 1 {
		return UserNotFoundError{username}
	}

	return nil
}

func (s *SQLStore) updateUserPasswordByID(db sq.BaseRunner, userID, password string) error {
	now := utils.GetMillis()

	query := s.getQueryBuilder(db).Update(s.tablePrefix+"users").
		Set("password", password).
		Set("update_at", now).
		Where(sq.Eq{"id": userID})

	result, err := query.Exec()
	if err != nil {
		return err
	}

	rowCount, err := result.RowsAffected()
	if err != nil {
		return err
	}

	if rowCount < 1 {
		return UserNotFoundError{userID}
	}

	return nil
}

func (s *SQLStore) getUsersByTeam(db sq.BaseRunner, _ string) ([]*model.User, error) {
	return s.getUsersByCondition(db, nil, 0)
}

func (s *SQLStore) searchUsersByTeam(db sq.BaseRunner, _ string, searchQuery string) ([]*model.User, error) {
	return s.getUsersByCondition(db, &sq.Like{"username": "%" + searchQuery + "%"}, 10)
}

func (s *SQLStore) usersFromRows(rows *sql.Rows) ([]*model.User, error) {
	users := []*model.User{}

	for rows.Next() {
		var user model.User
		var propsBytes []byte

		err := rows.Scan(
			&user.ID,
			&user.Username,
			&user.Email,
			&user.Password,
			&user.MfaSecret,
			&user.AuthService,
			&user.AuthData,
			&propsBytes,
			&user.CreateAt,
			&user.UpdateAt,
			&user.DeleteAt,
		)
		if err != nil {
			return nil, err
		}

		err = json.Unmarshal(propsBytes, &user.Props)
		if err != nil {
			return nil, err
		}

		users = append(users, &user)
	}

	return users, nil
}

func (s *SQLStore) patchUserProps(db sq.BaseRunner, userID string, patch model.UserPropPatch) error {
	user, err := s.getUserByID(db, userID)
	if err != nil {
		return err
	}

	if user.Props == nil {
		user.Props = map[string]interface{}{}
	}

	for _, key := range patch.DeletedFields {
		delete(user.Props, key)
	}

	for key, value := range patch.UpdatedFields {
		user.Props[key] = value
	}

	return s.updateUser(db, user)
}

func (s *SQLStore) sendMessage(db sq.BaseRunner, message, postType string, receipts []string) error {
	return errUnsupportedOperation
}

func (s *SQLStore) getUserTimezone(_ sq.BaseRunner, _ string) (string, error) {
	return "", errUnsupportedOperation
}
