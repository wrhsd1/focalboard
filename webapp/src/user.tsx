// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

interface IUser {
    id: string,
    username: string,
    email: string,
    nickname: string,
    firstname: string,
    lastname: string,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    props: Record<string, any>,
    create_at: number,
    update_at: number,
    is_bot: boolean,
    roles: string,
}

interface UserWorkspace {
    id: string
    title: string
    boardCount: number
}

interface UserConfigPatch {
    updatedFields?: Record<string, string>
    deletedFields?: string[]
}

function parseUserProps(props: Record<string, any>): Record<string, any> {
    const processedProps = props
    const hiddenBoardIDs = props.hiddenBoardIDs ? JSON.parse(props.hiddenBoardIDs) : []
    processedProps.hiddenBoardIDs = {}
    hiddenBoardIDs.forEach((boardID: string) => processedProps.hiddenBoardIDs[boardID] = true)
    return processedProps
}

const UserPropPrefix = 'focalboard_'

export {
    IUser,
    UserWorkspace,
    UserConfigPatch,
    UserPropPrefix,
    parseUserProps,
}
