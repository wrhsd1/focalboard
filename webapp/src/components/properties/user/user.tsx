// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react'
import Select from 'react-select'
import {CSSObject} from '@emotion/serialize'

import {Utils} from '../../../utils'

import {IUser} from '../../../user'

import {getBoardUsersList, getBoardUsers} from '../../../store/users'
import {useAppSelector} from '../../../store/hooks'

import './user.scss'
import {getSelectBaseStyle} from '../../../theme'
import {ClientConfig} from '../../../config/clientConfig'
import {getClientConfig} from '../../../store/clientConfig'
import {propertyValueClassName} from '../../propertyValueUtils'

const imageURLForUser = (window as any).Components?.imageURLForUser

type Props = {
    value: string,
    readonly: boolean,
    onChange: (value: string) => void,
}

const selectStyles = {
    ...getSelectBaseStyle(),
    option: (provided: CSSObject, state: {isFocused: boolean}): CSSObject => ({
        ...provided,
        background: state.isFocused ? 'rgba(var(--center-channel-color-rgb), 0.1)' : 'rgb(var(--center-channel-bg-rgb))',
        color: state.isFocused ? 'rgb(var(--center-channel-color-rgb))' : 'rgb(var(--center-channel-color-rgb))',
        padding: '8px',
    }),
    control: (): CSSObject => ({
        border: 0,
        width: '100%',
        margin: '0',
    }),
    valueContainer: (provided: CSSObject): CSSObject => ({
        ...provided,
        padding: 'unset',
        overflow: 'unset',
    }),
    singleValue: (provided: CSSObject): CSSObject => ({
        ...provided,
        position: 'static',
        top: 'unset',
        transform: 'unset',
    }),
    menu: (provided: CSSObject): CSSObject => ({
        ...provided,
        width: 'unset',
        background: 'rgb(var(--center-channel-bg-rgb))',
        minWidth: '260px',
    }),
}

const UserProperty = (props: Props): JSX.Element => {
    const clientConfig = useAppSelector<ClientConfig>(getClientConfig)

    const formatOptionLabel = (user: any) => {
        let profileImg
        if (imageURLForUser) {
            profileImg = imageURLForUser(user.id)
        }
    
        return (
            <div className='UserProperty-item'>
                {profileImg && (
                    <img
                        alt='UserProperty-avatar'
                        src={profileImg}
                    />
                )}
                {Utils.getUserDisplayName(user, clientConfig.teammateNameDisplay)}
            </div>
        )
    }

    const boardUsersById = useAppSelector<{[key:string]: IUser}>(getBoardUsers)

    const user = boardUsersById[props.value]

    if (props.readonly) {
        return (
            <div className={`UserProperty ${propertyValueClassName({readonly: true})}`}>
                {user ? formatOptionLabel(user) : props.value}
            </div>
        )
    }

    const boardUsers = useAppSelector<IUser[]>(getBoardUsersList)

    return (
        <Select
            options={boardUsers}
            isSearchable={true}
            isClearable={true}
            backspaceRemovesValue={true}
            className={`UserProperty ${propertyValueClassName()}`}
            classNamePrefix={'react-select'}
            formatOptionLabel={formatOptionLabel}
            styles={selectStyles}
            placeholder={'Empty'}
            getOptionLabel={(o: IUser) => o.username}
            getOptionValue={(a: IUser) => a.id}
            value={boardUsersById[props.value] || null}
            onChange={(item, action) => {
                if (action.action === 'select-option') {
                    props.onChange(item?.id || '')
                } else if (action.action === 'clear') {
                    props.onChange('')
                }
            }}
        />
    )
}

export default UserProperty
