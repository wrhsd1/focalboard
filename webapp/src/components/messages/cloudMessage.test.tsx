// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react'
import {Provider as ReduxProvider} from 'react-redux'

import {render, screen} from '@testing-library/react'
import {mocked} from 'jest-mock'
import userEvent from '@testing-library/user-event'

import configureStore from 'redux-mock-store'

import {Utils} from '../../utils'

import {IUser} from '../../user'

import {wrapIntl} from '../../testUtils'

import client from '../../octoClient'

import {UserSettings} from '../../userSettings'

import CloudMessage from './cloudMessage'

jest.mock('../../utils')
jest.mock('../../octoClient')
const mockedOctoClient = mocked(client, true)

describe('components/messages/CloudMessage', () => {
    beforeEach(() => {
        jest.clearAllMocks()
    })

    const mockedUtils = mocked(Utils, true)
    const mockStore = configureStore([])

    test('plugin mode, no display', () => {
        mockedUtils.isFocalboardPlugin.mockReturnValue(true)

        const me: IUser = {
            id: 'user-id-1',
            username: 'username_1',
            email: '',
            nickname: '',
            firstname: '', 
            lastname: '',
            props: {},
            create_at: 0,
            update_at: 0,
            is_bot: false,
            roles: 'system_user',
        }
        const state = {
            users: {
                me,
            },
        }

        const store = mockStore(state)

        const component = wrapIntl(
            <ReduxProvider store={store}>
                <CloudMessage/>
            </ReduxProvider>,
        )

        const {container} = render(component)
        expect(container).toMatchSnapshot()
    })

    test('not plugin mode, close message', () => {
        const me: IUser = {
            id: 'user-id-1',
            username: 'username_1',
            email: '',
            nickname: '',
            firstname: '', 
            lastname: '',
            props: {
                focalboard_cloudMessageCanceled: 'true',
            },
            create_at: 0,
            update_at: 0,
            is_bot: false,
            roles: 'system_user',
        }
        const state = {
            users: {
                me,
            },
        }
        const store = mockStore(state)
        mockedUtils.isFocalboardPlugin.mockReturnValue(false)

        const component = wrapIntl(
            <ReduxProvider store={store}>
                <CloudMessage/>
            </ReduxProvider>,
        )

        const {container} = render(component)
        expect(container).toMatchSnapshot()
    })

    test('not plugin mode, show message, close message', () => {
        const me: IUser = {
            id: 'user-id-1',
            username: 'username_1',
            email: '',
            nickname: '',
            firstname: '', 
            lastname: '',
            props: {},
            create_at: 0,
            update_at: 0,
            is_bot: false,
            roles: 'system_user',
        }
        const state = {
            users: {
                me,
            },
        }
        const store = mockStore(state)
        mockedUtils.isFocalboardPlugin.mockReturnValue(false)

        const component = wrapIntl(
            <ReduxProvider store={store}>
                <CloudMessage/>
            </ReduxProvider>,
        )

        const {container} = render(component)
        expect(container).toMatchSnapshot()

        const buttonElement = screen.getByRole('button', {name: 'Close dialog'})
        userEvent.click(buttonElement)
        expect(mockedOctoClient.patchUserConfig).toBeCalledWith('user-id-1', {
            updatedFields: {
                focalboard_cloudMessageCanceled: 'true',
            },
        })
    })

    test('not plugin mode, single user, close message', () => {
        const me: IUser = {
            id: 'single-user',
            username: 'single-user',
            email: 'single-user',
            nickname: '',
            firstname: '', 
            lastname: '',
            props: {},
            create_at: 0,
            update_at: Date.now() - (1000 * 60 * 60 * 24), //24 hours,
            is_bot: false,
            roles: 'system_user',
        }
        const state = {
            users: {
                me,
            },
        }
        const store = mockStore(state)
        const hideCloudMessageSpy = jest.spyOn(UserSettings, 'hideCloudMessage', 'set')

        mockedUtils.isFocalboardPlugin.mockReturnValue(false)

        const component = wrapIntl(
            <ReduxProvider store={store}>
                <CloudMessage/>
            </ReduxProvider>,
        )

        const {container} = render(component)
        expect(container).toMatchSnapshot()

        const buttonElement = screen.getByRole('button', {name: 'Close dialog'})
        userEvent.click(buttonElement)

        expect(mockedOctoClient.patchUserConfig).toBeCalledTimes(0)
        expect(hideCloudMessageSpy).toHaveBeenCalledWith(true)
        expect(UserSettings.hideCloudMessage).toBe(true)
    })
})
