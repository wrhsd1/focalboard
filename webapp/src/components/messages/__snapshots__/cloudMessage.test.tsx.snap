// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/messages/CloudMessage not plugin mode, close message 1`] = `<div />`;

exports[`components/messages/CloudMessage not plugin mode, show message, close message 1`] = `
<div>
  <div
    class="CloudMessage"
  >
    <div
      class="banner"
    >
      <i
        class="CompassIcon icon-information-outline CompassIcon"
      />
      Get your own free cloud server.
      <button
        title="Learn more"
        type="button"
      >
        <span>
          Learn more
        </span>
      </button>
    </div>
    <button
      aria-label="Close dialog"
      title="Close dialog"
      type="button"
    >
      <i
        class="CompassIcon icon-close CloseIcon"
      />
    </button>
  </div>
</div>
`;

exports[`components/messages/CloudMessage not plugin mode, single user, close message 1`] = `
<div>
  <div
    class="CloudMessage"
  >
    <div
      class="banner"
    >
      <i
        class="CompassIcon icon-information-outline CompassIcon"
      />
      Get your own free cloud server.
      <button
        title="Learn more"
        type="button"
      >
        <span>
          Learn more
        </span>
      </button>
    </div>
    <button
      aria-label="Close dialog"
      title="Close dialog"
      type="button"
    >
      <i
        class="CompassIcon icon-close CloseIcon"
      />
    </button>
  </div>
</div>
`;

exports[`components/messages/CloudMessage plugin mode, no display 1`] = `<div />`;
